client/node-common/src/service.rs

```diff
@@ -0,0 +1,869 @@
+    /// stores the returned objects in `self.network` and `self.tx_handler_controller`.
+    ///
+    /// Can only be called once on a `NodeBuilder` that doesn't have yet network
+    /// data.
+    #[must_use]
+    pub async fn build_cumulus_network<RCInterface>(
+        self,
+        parachain_config: &Configuration,
+        para_id: ParaId,
+        import_queue: impl ImportQueue<BlockOf<T>> + 'static,
+        relay_chain_interface: RCInterface,
+    ) -> sc_service::error::Result<
+        NodeBuilder<
+            T,
+            Network<BlockOf<T>>,
+            TransactionsHandlerController<BlockHashOf<T>>,
+            ImportQueueServiceOf<T>,
+        >,
+    >
+    where
+        SNetwork: TypeIdentity<Type = ()>,
+        STxHandler: TypeIdentity<Type = ()>,
+        SImportQueueService: TypeIdentity<Type = ()>,
+        RCInterface: RelayChainInterface + Clone + 'static,
+    {
+        let Self {
+            client,
+            backend,
+            transaction_pool,
+            telemetry,
+            telemetry_worker_handle,
+            task_manager,
+            keystore_container,
+            hwbench,
+            prometheus_registry,
+            network: _,
+            tx_handler_controller: _,
+            import_queue_service: _,
+        } = self;
+
+        let net_config = FullNetworkConfiguration::new(&parachain_config.network);
+        let import_queue_service = import_queue.service();
+
+        let (network, system_rpc_tx, tx_handler_controller, start_network, sync_service) =
+            cumulus_client_service::build_network(cumulus_client_service::BuildNetworkParams {
+                parachain_config: &parachain_config,
+                client: client.clone(),
+                transaction_pool: transaction_pool.clone(),
+                spawn_handle: task_manager.spawn_handle(),
+                import_queue: import_queue,
+                para_id,
+                relay_chain_interface: relay_chain_interface,
+                net_config,
+                sybil_resistance_level: CollatorSybilResistance::Resistant,
+            })
+            .await?;
+
+        Ok(NodeBuilder {
+            client,
+            backend,
+            transaction_pool,
+            telemetry,
+            telemetry_worker_handle,
+            task_manager,
+            keystore_container,
+            hwbench,
+            prometheus_registry,
+            network: Network {
+                network,
+                system_rpc_tx,
+                start_network,
+                sync_service,
+            },
+            tx_handler_controller,
+            import_queue_service,
+        })
+    }
+
+    /// Given an import queue, calls `sc_service::build_network` and
+    /// stores the returned objects in `self.network` and `self.tx_handler_controller`.
+    ///
+    /// Can only be called once on a `NodeBuilder` that doesn't have yet network
+    /// data.
+    #[must_use]
+    pub fn build_substrate_network(
```

Comment:
So I cannot call together both `build_substrate_network` and `build_cumulus_network` right? I wonder whether this can be an enum as an argument, where in the case of type:

```
pub struct CumulusNetwork {
para_id: ParaId,
relay_chain_interface: RelayChainInterface
}
pub enum NetworkType {
Substrate,
Cumulus(CumulusNetwork)
}
```

Reply:
This enum would need to be generic over the RelayChainInterface type, which means that in Substrate mode it will not be able to infer the type and will require to explicitly name one.